# Lab 8: Define and Solve an ML Problem of Your Choosing

In [1]:
import pandas as pd
import numpy as np
import os 
import matplotlib.pyplot as plt
import seaborn as sns

In this lab assignment, you will follow the machine learning life cycle and implement a model to solve a machine learning problem of your choosing. You will select a data set and choose a predictive problem that the data set supports.  You will then inspect the data with your problem in mind and begin to formulate a  project plan. You will then implement the machine learning project plan. 

You will complete the following tasks:

1. Build Your DataFrame
2. Define Your ML Problem
3. Perform exploratory data analysis to understand your data.
4. Define Your Project Plan
5. Implement Your Project Plan:
    * Prepare your data for your model.
    * Fit your model to the training data and evaluate your model.
    * Improve your model's performance.

## Part 1: Build Your DataFrame

You will have the option to choose one of four data sets that you have worked with in this program:

* The "census" data set that contains Census information from 1994: `censusData.csv`
* Airbnb NYC "listings" data set: `airbnbListingsData.csv`
* World Happiness Report (WHR) data set: `WHR2018Chapter2OnlineData.csv`
* Book Review data set: `bookReviewsData.csv`

Note that these are variations of the data sets that you have worked with in this program. For example, some do not include some of the preprocessing necessary for specific models. 

#### Load a Data Set and Save it as a Pandas DataFrame

The code cell below contains filenames (path + filename) for each of the four data sets available to you.

<b>Task:</b> In the code cell below, use the same method you have been using to load the data using `pd.read_csv()` and save it to DataFrame `df`. 

You can load each file as a new DataFrame to inspect the data before choosing your data set.

In [2]:
# File names of the four data sets
adultDataSet_filename = os.path.join(os.getcwd(), "data", "censusData.csv")
airbnbDataSet_filename = os.path.join(os.getcwd(), "data", "airbnbListingsData.csv")
WHRDataSet_filename = os.path.join(os.getcwd(), "data", "WHR2018Chapter2OnlineData.csv")
bookReviewDataSet_filename = os.path.join(os.getcwd(), "data", "bookReviewsData.csv")


df = pd.read_csv(airbnbDataSet_filename)

df.head()

name  \
0                              Skylit Midtown Castle   
1  Whole flr w/private bdrm, bath & kitchen(pls r...   
2           Spacious Brooklyn Duplex, Patio + Garden   
3                   Large Furnished Room Near B'way　   
4                 Cozy Clean Guest Room - Family Apt   

                                         description  \
0  Beautiful, spacious skylit studio in the heart...   
1  Enjoy 500 s.f. top floor in 1899 brownstone, w...   
2  We welcome you to stay in our lovely 2 br dupl...   
3  Please don’t expect the luxury here just a bas...   
4  Our best guests are seeking a safe, clean, spa...   

                               neighborhood_overview    host_name  \
0  Centrally located in the heart of Manhattan ju...     Jennifer   
1  Just the right mix of urban center and local n...  LisaRoxanne   
2                                                NaN      Rebecca   
3    Theater district, many restaurants around here.     Shunichi   
4  Our neighborhood is full of restaurants and ca...    MaryEllen   

                       host_location  \
0  New York, New York, United States   
1  New York, New York, United States   
2  Brooklyn, New York, United States   
3  New York, New York, United States   
4  New York, New York, United States   

                                          host_about  host_response_rate  \
0  A New Yorker since 2000! My passion is creatin...                0.80   
1  Laid-back Native New Yorker (formerly bi-coast...                0.09   
2  Rebecca is an artist/designer, and Henoch is i...                1.00   
3  I used to work for a financial industry but no...                1.00   
4  Welcome to family life with my oldest two away...                 NaN   

   host_acceptance_rate  host_is_superhost  host_listings_count  ...  \
0                  0.17               True                  8.0  ...   
1                  0.69               True                  1.0  ...   
2                  0.25               True                  1.0  ...   
3                  1.00               True                  1.0  ...   
4                   NaN               True                  1.0  ...   

   review_scores_communication  review_scores_location  review_scores_value  \
0                         4.79                    4.86                 4.41   
1                         4.80                    4.71                 4.64   
2                         5.00                    4.50                 5.00   
3                         4.42                    4.87                 4.36   
4                         4.95                    4.94                 4.92   

  instant_bookable calculated_host_listings_count  \
0            False                              3   
1            False                              1   
2            False                              1   
3            False                              1   
4            False                              1   

   calculated_host_listings_count_entire_homes  \
0                                            3   
1                                            1   
2                                            1   
3                                            0   
4                                            0   

   calculated_host_listings_count_private_rooms  \
0                                             0   
1                                             0   
2                                             0   
3                                             1   
4                                             1   

   calculated_host_listings_count_shared_rooms  reviews_per_month  \
0                                            0               0.33   
1                                            0               4.86   
2                                            0               0.02   
3                                            0               3.68   
4                                            0               0.87   

  n_host_ve

## Part 2: Define Your ML Problem

Next you will formulate your ML Problem. In the markdown cell below, answer the following questions:

1. List the data set you have chosen.
2. What will you be predicting? What is the label?
3. Is this a supervised or unsupervised learning problem? Is this a clustering, classification or regression problem? Is it a binary classificaiton or multi-class classifiction problem?
4. What are your features? (note: this list may change after your explore your data)
5. Explain why this is an important problem. In other words, how would a company create value with a model that predicts this label?

I have chosen the airbnb dataset which will be predicting the price of a certain listing, making the label would be the price coulmn. It will be a supervised learning problem, specifically a regression task, since we are predicting a continuous value. Some important features that I'll include would be room_type, accommodates, bathrooms, bedrooms, beds, minimum_nights, availability_365, number_of_reviews, and review_scores_rating. Predicting rental prices is valuable because it helps the hosts support flexible pricing and provide insights for market analysis and cost management. I would consider this an important problem because accurate predictions can boost host cost revenue and enhance user satisfaction by offering better pricing information.

## Part 3: Understand Your Data

The next step is to perform exploratory data analysis. Inspect and analyze your data set with your machine learning problem in mind. Consider the following as you inspect your data:

1. What data preparation techniques would you like to use? These data preparation techniques may include:

    * addressing missingness, such as replacing missing values with means
    * finding and replacing outliers
    * renaming features and labels
    * finding and replacing outliers
    * performing feature engineering techniques such as one-hot encoding on categorical features
    * selecting appropriate features and removing irrelevant features
    * performing specific data cleaning and preprocessing techniques for an NLP problem
    * addressing class imbalance in your data sample to promote fair AI
    

2. What machine learning model (or models) you would like to use that is suitable for your predictive problem and data?
    * Are there other data preparation techniques that you will need to apply to build a balanced modeling data set for your problem and model? For example, will you need to scale your data?
 
 
3. How will you evaluate and improve the model's performance?
    * Are there specific evaluation metrics and methods that are appropriate for your model?
    

Think of the different techniques you have used to inspect and analyze your data in this course. These include using Pandas to apply data filters, using the Pandas `describe()` method to get insight into key statistics for each column, using the Pandas `dtypes` property to inspect the data type of each column, and using Matplotlib and Seaborn to detect outliers and visualize relationships between features and labels. If you are working on a classification problem, use techniques you have learned to determine if there is class imbalance.

<b>Task</b>: Use the techniques you have learned in this course to inspect and analyze your data. You can import additional packages that you have used in this course that you will need to perform this task.

<b>Note</b>: You can add code cells if needed by going to the <b>Insert</b> menu and clicking on <b>Insert Cell Below</b> in the drop-drown menu.

In [3]:
# YOUR CODE HERE
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from scipy import stats

In [4]:
# handling missing values
df['review_scores_rating'].fillna(df['review_scores_rating'].mean(), inplace=True)
df['room_type'].fillna('Unknown', inplace=True)

# handeling outliers
df['price'] = np.log1p(df['price'])  

# feature engineering 
df = pd.get_dummies(df, columns=['room_type'])

# feaature selection
df.drop(columns=['host_name', 'host_location'], inplace=True)

# scaling numerical features
features = ['accommodates', 'bathrooms', 'bedrooms', 'beds', 'minimum_nights', 'availability_365', 'number_of_reviews', 'review_scores_rating']
X = df[features]
y = df['price']
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# train test split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# model training
model = RandomForestRegressor()
model.fit(X_train, y_train)

# predications and evaluation
y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f'MAE: {mae}')
print(f'MSE: {mse}')
print(f'R^2: {r2}')


MAE: 0.4012916418948819
MSE: 0.2679681253881403
R^2: 0.43515844627293965


## Part 4: Define Your Project Plan

Now that you understand your data, in the markdown cell below, define your plan to implement the remaining phases of the machine learning life cycle (data preparation, modeling, evaluation) to solve your ML problem. Answer the following questions:

* Do you have a new feature list? If so, what are the features that you chose to keep and remove after inspecting the data? 
* Explain different data preparation techniques that you will use to prepare your data for modeling.
* What is your model (or models)?
* Describe your plan to train your model, analyze its performance and then improve the model. That is, describe your model building, validation and selection plan to produce a model that generalizes well to new data. 

The project plan is to keep features like accommodates, bathrooms, bedrooms, and review_scores_rating, while removing host_name and host_location. Data preparation will include filling missing values, handling outliers, one hot encoding categorical variables, and scaling numerical features. We'll use a Random Forest Regressor for modeling and the model will be trained on the training set, evaluated using MAE, MSE, and R² on the test set, and improved through tuning and cross-validation.

## Part 5: Implement Your Project Plan

<b>Task:</b> In the code cell below, import additional packages that you have used in this course that you will need to implement your project plan.

In [5]:
# YOUR CODE HERE
from sklearn.model_selection import GridSearchCV

<b>Task:</b> Use the rest of this notebook to carry out your project plan. 

You will:

1. Prepare your data for your model.
2. Fit your model to the training data and evaluate your model.
3. Improve your model's performance by performing model selection and/or feature selection techniques to find best model for your problem.

Add code cells below and populate the notebook with commentary, code, analyses, results, and figures as you see fit. 

In [6]:
# YOUR CODE HERE
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [10, 20, 30],
    'min_samples_split': [2, 5, 10]
}

grid_search = GridSearchCV(estimator=RandomForestRegressor(random_state=42),
                           param_grid=param_grid,
                           cv=5,
                           scoring='neg_mean_squared_error',
                           n_jobs=-1,
                           verbose=2)
grid_search.fit(X_train, y_train)

# the BEST parameters and model
best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

# predict with the bestt model
y_pred_best = best_model.predict(X_test)

# evaluate the improved model
mae_best = mean_absolute_error(y_test, y_pred_best)
mse_best = mean_squared_error(y_test, y_pred_best)
r2_best = r2_score(y_test, y_pred_best)

print(f'best MAE: {mae_best}')
print(f'best MSE: {mse_best}')
print(f'best R^2: {r2_best}')

Fitting 5 folds for each of 27 candidates, totalling 135 fits
best MAE: 0.39247872261795774
best MSE: 0.2512385335840639
best R^2: 0.47042222480684825
[CV] END max_depth=10, min_samples_split=2, n_estimators=100; total time=   2.7s
[CV] END max_depth=10, min_samples_split=2, n_estimators=200; total time=   5.3s
[CV] END max_depth=10, min_samples_split=5, n_estimators=100; total time=   2.6s
[CV] END max_depth=10, min_samples_split=5, n_estimators=100; total time=   2.6s
[CV] END max_depth=10, min_samples_split=5, n_estimators=300; total time=   7.9s
[CV] END max_depth=10, min_samples_split=10, n_estimators=200; total time=   5.2s
[CV] END max_depth=10, min_samples_split=10, n_estimators=300; total time=   7.8s
[CV] END max_depth=20, min_samples_split=2, n_estimators=300; total time=  14.1s
[CV] END max_depth=20, min_samples_split=5, n_estimators=200; total time=   8.5s
[CV] END max_depth=20, min_samples_split=10, n_estimators=100; total time=   3.9s
[CV] END max_depth=20, min_samples_s

The model being tested with 27 different sets of parameters, and for each set, it's evaluated 5 times on different subsets of the data. In total, this results in 135 separate evaluations of the model's performance, and the best model achieved a mean absolute error of 0.392, a mean squared error (MSE) of 0.251, and an R² score of 0.470, indicating its best performance.